File: examples/by-algorithm/lc0155/go/lc0155_min_stack.go
package leetcode

// LeetCode 155. Min Stack
// 最小栈
//
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

// MinStack 主栈 + 辅助栈同步维护最小值
// Main stack + auxiliary stack maintaining minimum
type MinStack struct {
	stack    []int
	minStack []int
}

// ConstructorMinStack initializes the stack object
func ConstructorMinStack() MinStack {
	return MinStack{stack: []int{}, minStack: []int{}}
}

// Push pushes the element val onto the stack
func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)
	if len(this.minStack) == 0 {
		this.minStack = append(this.minStack, val)
	} else {
		minVal := val
		if this.minStack[len(this.minStack)-1] < minVal {
			minVal = this.minStack[len(this.minStack)-1]
		}
		this.minStack = append(this.minStack, minVal)
	}
}

// Pop removes the element on the top of the stack
func (this *MinStack) Pop() {
	this.stack = this.stack[:len(this.stack)-1]
	this.minStack = this.minStack[:len(this.minStack)-1]
}

// Top gets the top element of the stack
func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1]
}

// GetMin retrieves the minimum element in the stack
func (this *MinStack) GetMin() int {
	return this.minStack[len(this.minStack)-1]
}

File: examples/by-algorithm/lc0155/rust/lc0155_min_stack.rs
//! LeetCode 155. Min Stack
//! 最小栈
//!
//! Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

pub struct MinStack {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

impl MinStack {
    /// 初始化栈对象
    /// Initialize the stack object
    pub fn new() -> Self {
        MinStack {
            stack: Vec::new(),
            min_stack: Vec::new(),
        }
    }

    /// 将元素压入栈
    /// Push element onto stack
    pub fn push(&mut self, val: i32) {
        self.stack.push(val);
        let min_val = self.min_stack.last().map_or(val, |&m| m.min(val));
        self.min_stack.push(min_val);
    }

    /// 移除栈顶元素
    /// Remove the element on the top of the stack
    pub fn pop(&mut self) {
        self.stack.pop();
        self.min_stack.pop();
    }

    /// 获取栈顶元素
    /// Get the top element
    pub fn top(&self) -> i32 {
        *self.stack.last().unwrap()
    }

    /// 获取最小元素
    /// Retrieve the minimum element
    pub fn get_min(&self) -> i32 {
        *self.min_stack.last().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_min_stack() {
        let mut ms = MinStack::new();
        ms.push(-2);
        ms.push(0);
        ms.push(-3);
        assert_eq!(ms.get_min(), -3);
        ms.pop();
        assert_eq!(ms.top(), 0);
        assert_eq!(ms.get_min(), -2);
    }
}
